File: packages/domain-types/src/events/enrichment.ts
/**
 * Job Enrichment domain events.
 *
 * @see docs/architecture/domain-model/tactical.md §4.2
 *
 * PR3 additions: PostingContentSnapshotCaptured, PostingContentSnapshotFailed,
 * JobActiveStateChanged, ContentDuplicateCandidateDetected. Each event shape
 * mirrors the §"Domain Events" table in
 * docs/plans/implemented/2026-05-12-job-search-discovery-rfc.md.
 */

import type { TenantId } from "../tenant.js";
import { type DomainEvent, createDomainEvent } from "./base.js";

// -- JobEnriched ------------------------------------------------------------

export interface JobEnrichedPayload {
  readonly jobId: string;
  readonly fullDescription: string;
  readonly applicationUrl: string;
  readonly extractionTier: string;
  readonly enrichedAt: string;
}

export type JobEnriched = DomainEvent<"JobEnriched", JobEnrichedPayload>;

export function createJobEnriched(
  tenantId: TenantId,
  payload: JobEnrichedPayload,
): JobEnriched {
  return createDomainEvent("JobEnriched", tenantId, payload);
}

// -- EnrichmentFailed -------------------------------------------------------

export interface EnrichmentFailedPayload {
  readonly jobId: string;
  readonly error: string;
  readonly attemptNumber: number;
}

export type EnrichmentFailed = DomainEvent<"EnrichmentFailed", EnrichmentFailedPayload>;

export function createEnrichmentFailed(
  tenantId: TenantId,
  payload: EnrichmentFailedPayload,
): EnrichmentFailed {
  return createDomainEvent("EnrichmentFailed", tenantId, payload);
}

// -- ActiveState ------------------------------------------------------------
// Mirrors workers/automation/src/jobhunter/domain/enrichment/snapshot_value_objects.py
// `ActiveState`. Kept as a string-literal union so JSON-RPC payloads can carry
// the same shape without a separate enum import path.

export type ActiveStateValue =
  | "unknown"
  | "active"
  | "closed"
  | "expired"
  | "removed"
  | "location_incompatible";

// -- DuplicateEvidenceKind --------------------------------------------------
// Mirrors workers/automation/src/jobhunter/domain/enrichment/snapshot_value_objects.py
// `DuplicateEvidenceKind` so the SSE payload shape matches the Python writer.

export type DuplicateEvidenceKindValue =
  | "description_hash_match"
  | "apply_url_match"
  | "high_confidence_content_similarity";

// -- DuplicateEvidence on the wire -----------------------------------------

export interface DuplicateEvidenceOnWire {
  readonly kind: DuplicateEvidenceKindValue;
  readonly matchedValue: string;
  readonly confidence: number;
}

// -- PostingContentSnapshotCaptured ----------------------------------------

export interface PostingContentSnapshotCapturedPayload {
  readonly jobId: string;
  readonly snapshotVersion: number;
  readonly snapshotRef: string;
  readonly sourceId: string;
  readonly extractionTier: string;
  readonly capturedAt: string;
}

export type PostingContentSnapshotCaptured = DomainEvent<
  "PostingContentSnapshotCaptured",
  PostingContentSnapshotCapturedPayload
>;

export function createPostingContentSnapshotCaptured(
  tenantId: TenantId,
  payload: PostingContentSnapshotCapturedPayload,
): PostingContentSnapshotCaptured {
  return createDomainEvent("PostingContentSnapshotCaptured", tenantId, payload);
}

// -- PostingContentSnapshotFailed ------------------------------------------

export interface PostingContentSnapshotFailedPayload {
  readonly jobId: string;
  readonly sourceId: string;
  readonly errorClass: string;
  readonly retryable: boolean;
  readonly failedAt: string;
}

export type PostingContentSnapshotFailed = DomainEvent<
  "PostingContentSnapshotFailed",
  PostingContentSnapshotFailedPayload
>;

export function createPostingContentSnapshotFailed(
  tenantId: TenantId,
  payload: PostingContentSnapshotFailedPayload,
): PostingContentSnapshotFailed {
  return createDomainEvent("PostingContentSnapshotFailed", tenantId, payload);
}

// -- JobActiveStateChanged --------------------------------------------------

export interface JobActiveStateChangedPayload {
  readonly jobId: string;
  readonly activeState: ActiveStateValue;
  readonly previousState: ActiveStateValue;
  readonly verificationMethod: string;
  readonly verifiedAt: string;
}

export type JobActiveStateChanged = DomainEvent<
  "JobActiveStateChanged",
  JobActiveStateChangedPayload
>;

export function createJobActiveStateChanged(
  tenantId: TenantId,
  payload: JobActiveStateChangedPayload,
): JobActiveStateChanged {
  return createDomainEvent("JobActiveStateChanged", tenantId, payload);
}

// -- ContentDuplicateCandidateDetected -------------------------------------

export interface ContentDuplicateCandidateDetectedPayload {
  readonly jobId: string;
  readonly candidateJobId: string;
  readonly evidence: readonly DuplicateEvidenceOnWire[];
  readonly confidence: number;
  readonly detectedAt: string;
}

export type ContentDuplicateCandidateDetected = DomainEvent<
  "ContentDuplicateCandidateDetected",
  ContentDuplicateCandidateDetectedPayload
>;

export function createContentDuplicateCandidateDetected(
  tenantId: TenantId,
  payload: ContentDuplicateCandidateDetectedPayload,
): ContentDuplicateCandidateDetected {
  return createDomainEvent("ContentDuplicateCandidateDetected", tenantId, payload);
}

File: workers/automation/src/jobhunter/domain/events/enrichment.py
"""Job Enrichment domain events.

See ddd-target.md §4.2.

PR3 additions: ``PostingContentSnapshotCaptured``,
``PostingContentSnapshotFailed``, ``JobActiveStateChanged``,
``ContentDuplicateCandidateDetected``. Each event shape mirrors the
§"Domain Events" table in
``docs/plans/proposed/2026-05-12-job-search-discovery-rfc.md`` so the
Operations projections can stay exhaustive and the SSE invalidation
router can stay parity-safe.
"""

from __future__ import annotations

from dataclasses import asdict, dataclass, field
from typing import Any

from jobhunter.domain.tenant import TenantId
from jobhunter.domain.events.base import DomainEvent, create_domain_event


@dataclass(frozen=True)
class JobEnrichedPayload:
    job_id: str
    full_description: str
    application_url: str
    extraction_tier: str
    enriched_at: str


def create_job_enriched(tenant_id: TenantId, payload: JobEnrichedPayload) -> DomainEvent:
    return create_domain_event("JobEnriched", tenant_id, asdict(payload))


@dataclass(frozen=True)
class EnrichmentFailedPayload:
    job_id: str
    error: str
    attempt_number: int


def create_enrichment_failed(tenant_id: TenantId, payload: EnrichmentFailedPayload) -> DomainEvent:
    return create_domain_event("EnrichmentFailed", tenant_id, asdict(payload))


# -- PR 3 events: PostingSnapshotSet + active verification + content dedupe -


@dataclass(frozen=True)
class PostingContentSnapshotCapturedPayload:
    """A new ``PostingContentSnapshot`` was written to the aggregate.

    Carries the snapshot's identity / provenance so Operations can
    update source-quality projections and the dedupe service can
    react on the description-hash join. ``snapshot_ref`` is a stable
    reference into the aggregate (``"<jobId>:<snapshotVersion>"``).
    """

    job_id: str
    snapshot_version: int
    snapshot_ref: str
    source_id: str
    extraction_tier: str
    captured_at: str


def create_posting_content_snapshot_captured(
    tenant_id: TenantId,
    payload: PostingContentSnapshotCapturedPayload,
) -> DomainEvent:
    return create_domain_event(
        "PostingContentSnapshotCaptured", tenant_id, asdict(payload)
    )


@dataclass(frozen=True)
class PostingContentSnapshotFailedPayload:
    """A snapshot capture attempt failed without writing a snapshot.

    ``error_class`` is the coarse classification used by source
    quality and the scheduler retry policy (``timeout`` / ``http_4xx``
    / ``parse_error`` etc.). The full error message stays out of the
    payload to keep the event small and observable-safe.
    """

    job_id: str
    source_id: str
    error_class: str
    retryable: bool
    failed_at: str


def create_posting_content_snapshot_failed(
    tenant_id: TenantId,
    payload: PostingContentSnapshotFailedPayload,
) -> DomainEvent:
    return create_domain_event(
        "PostingContentSnapshotFailed", tenant_id, asdict(payload)
    )


@dataclass(frozen=True)
class JobActiveStateChangedPayload:
    """The Enrichment-owned active-state for a Job changed.

    Emitted whenever the active verifier observes a transition (e.g.
    ``unknown`` → ``active``, ``active`` → ``closed``). The previous
    value is included so consumers can record audits without keeping
    their own state.
    """

    job_id: str
    active_state: str
    previous_state: str
    verification_method: str
    verified_at: str


def create_job_active_state_changed(
    tenant_id: TenantId,
    payload: JobActiveStateChangedPayload,
) -> DomainEvent:
    return create_domain_event("JobActiveStateChanged", tenant_id, asdict(payload))


@dataclass(frozen=True)
class ContentDuplicateCandidateDetectedPayload:
    """A new content-duplicate candidate was registered on the aggregate.

    Discovery confirms or rejects the link via ``DuplicateJobLink``;
    this event lets Operations show the queue without listening to
    Discovery's reaction. ``evidence`` is a list of dict rows so the
    payload stays serialisable.
    """

    job_id: str
    candidate_job_id: str
    evidence: list[dict[str, Any]] = field(default_factory=list)
    confidence: float = 0.0
    detected_at: str = ""


def create_content_duplicate_candidate_detected(
    tenant_id: TenantId,
    payload: ContentDuplicateCandidateDetectedPayload,
) -> DomainEvent:
    return create_domain_event(
        "ContentDuplicateCandidateDetected", tenant_id, asdict(payload)
    )
